File: ui/apps/web/src/lib/api/stream.ts
/** Start, and stop, an ephemeral on-demand HLS session for a site page, media URL or torrent. */

import { deleteApi, getApi, postApi } from "./client";

export type StreamSession = {
    sessionId: string;
    playlistUrl: string;
    status: string;
    durationSeconds: number | null;
    hasVideo: boolean | null;
};

/** One `stream_status` SSE payload — peer/speed/progress fields are only
 * present on torrent-backed sessions, and keep arriving through playback. */
export type StreamStatusEvent = {
    id: string;
    status: string;
    message?: string;
    peersConnected?: number;
    downloadBps?: number;
    progressBytes?: number;
    totalBytes?: number;
};

export function normalizeStreamStatusEvent(raw: any): StreamStatusEvent {
    return {
        id: raw?.id ?? "",
        status: raw?.status ?? "",
        message: raw?.message,
        peersConnected: raw?.peers_connected,
        downloadBps: raw?.download_bps,
        progressBytes: raw?.progress_bytes,
        totalBytes: raw?.total_bytes,
    };
}

export function normalizeStreamSession(raw: any): StreamSession {
    return {
        sessionId: raw?.session_id ?? "",
        playlistUrl: raw?.playlist_url ?? "",
        status: raw?.status ?? "ready",
        durationSeconds:
            raw?.duration_seconds ?? (raw?.status === "connecting" ? null : 0),
        hasVideo:
            raw?.has_video ?? (raw?.status === "connecting" ? null : false),
    };
}

export function isTorrentKind(kind: string): boolean {
    return kind === "magnet" || kind === "torrent";
}

export function buildStreamStartBody(
    value: string,
    kind: string,
): { url: string } | { torrent: string } {
    return isTorrentKind(kind) ? { torrent: value } : { url: value };
}

export async function startStream(
    value: string,
    kind: string,
): Promise<StreamSession> {
    return normalizeStreamSession(
        await postApi<any>("/stream/start", buildStreamStartBody(value, kind)),
    );
}

export function stopStream(sessionId: string): Promise<void> {
    return deleteApi(`/stream/${sessionId}`);
}

/** A finished download's file, as the player needs it to choose how to play it (#94). */
export type MediaInfo = {
    /** A torrent's file; `null` for a download's one file. */
    fileIndex: number | null;
    filename: string;
    durationSeconds: number;
    hasVideo: boolean;
    /** What `canPlayType` is asked about; `null` when no browser plays it from a file. */
    mediaType: string | null;
    /** The file itself, served with Range. */
    fileUrl: string;
};

export function normalizeMediaInfo(raw: any): MediaInfo {
    return {
        // The envelope leaves out what is null.
        fileIndex: raw?.file_index ?? null,
        filename: raw?.filename ?? "",
        durationSeconds: raw?.duration_seconds ?? 0,
        hasVideo: raw?.has_video ?? true,
        mediaType: raw?.media_type ?? null,
        fileUrl: raw?.file_url ?? "",
    };
}

export async function fetchMediaInfo(
    taskId: string,
    fileIndex: number | null,
): Promise<MediaInfo> {
    const query = fileIndex === null ? "" : `?file_index=${fileIndex}`;
    return normalizeMediaInfo(
        await getApi<any>(`/download/${taskId}/media${query}`),
    );
}

/**
 * The file itself when the browser says it can play it, else a session. Any
 * answer but "" counts: #93 found Chrome's exact and WebKit's too eager, and
 * `nativeFailed` catches the second.
 */
export function choosePlayback(
    media: Pick<MediaInfo, "mediaType">,
    canPlayType: (type: string) => string,
): "native" | "session" {
    return media.mediaType && canPlayType(media.mediaType) !== ""
        ? "native"
        : "session";
}

/**
 * Whether the file failed to play itself, so a session should take over:
 * the element errored, or it has loaded and shows no picture it should have.
 * WebKit answers "probably" to VP9 and then draws nothing (#93).
 */
export function nativeFailed(state: {
    errored: boolean;
    hasVideo: boolean;
    videoWidth: number;
}): boolean {
    return state.errored || (state.hasVideo && state.videoWidth === 0);
}

export function buildTaskStreamBody(
    taskId: string,
    fileIndex: number | null,
): { task_id: string; file_index?: number } {
    return fileIndex === null
        ? { task_id: taskId }
        : { task_id: taskId, file_index: fileIndex };
}

/** A session that reads a finished download's file from disk. */
export async function startTaskStream(
    taskId: string,
    fileIndex: number | null,
): Promise<StreamSession> {
    return normalizeStreamSession(
        await postApi<any>(
            "/stream/start",
            buildTaskStreamBody(taskId, fileIndex),
        ),
    );
}

File: api/src/data/schema/stream/stream.py
from __future__ import annotations

import uuid
from typing import Annotated

from pydantic import Field, model_validator

from src.core.base import BaseSchema


class StreamStartRequest(BaseSchema):
    url: Annotated[
        str | None,
        Field(
            default=None,
            min_length=1,
            description=(
                "A page on a site yt-dlp supports, played at up to 1080p, "
                "or a direct http or https URL to a media file"
            ),
        ),
    ]
    torrent: Annotated[
        str | None,
        Field(
            default=None,
            min_length=1,
            description="A magnet link, an http URL to a .torrent, or a base64 .torrent file",
        ),
    ]
    task_id: Annotated[
        uuid.UUID | None,
        Field(default=None, description="A finished download, played from its file on disk"),
    ]
    file_index: Annotated[
        int | None,
        Field(
            default=None,
            ge=0,
            description="With task_id, which of a torrent's files; its largest media file by default",
        ),
    ]

    @model_validator(mode="after")
    def _one_source(self) -> StreamStartRequest:
        sources = [self.url, self.torrent, self.task_id]
        if sum(source is not None for source in sources) != 1:
            raise ValueError("Provide exactly one of url, torrent or task_id")
        if self.file_index is not None and self.task_id is None:
            raise ValueError("file_index goes with task_id")
        return self


class MediaInfoSchema(BaseSchema):
    """What the player needs to choose between a download's file and a session (#94)."""

    #: A torrent's file; ``None`` for a download's one file.
    file_index: int | None = None
    filename: str
    duration_seconds: float
    has_video: bool
    #: For ``canPlayType``: ``None`` when no browser plays it from a file.
    media_type: str | None = None
    #: Where the browser fetches the file itself, with Range.
    file_url: str


class StreamSessionSchema(BaseSchema):
    session_id: str
    playlist_url: str
    status: str
    duration_seconds: float | None = None
    has_video: bool | None = None
